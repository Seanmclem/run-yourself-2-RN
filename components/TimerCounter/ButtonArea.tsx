import * as React from 'react';
import { StyleSheet } from 'react-native';
import { View } from '../../components/Themed';
import { AppButton } from '../AppButton';
import { Spacer } from '../Spacer';

interface props {
    isRunning: boolean;
    startTimer: any;
    nextLap: any;
    stopTimer: any;
}

export const ButtonArea: React.FC<props> = ({ isRunning, startTimer, nextLap, stopTimer }) => {
    return (
        <View style={styles.container}>
            { !isRunning && (
                <>
                    <AppButton
                        onPress={startTimer}
                        title="Start Timer"
                    />
                </>
            )}
            { isRunning && (
                <>
                    <AppButton
                        onPress={nextLap}
                        title="LAP"
                    />
                    <Spacer
                        width={50}
                        height={20}
                    />
                    <AppButton
                        onPress={stopTimer}
                        title="Stop Timer"
                    />
                </>
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
        justifyContent: 'center',
    },
});
