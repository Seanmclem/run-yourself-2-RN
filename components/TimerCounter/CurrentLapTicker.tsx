import * as React from 'react';
import { StyleSheet } from 'react-native';
import { Text, View } from '../../components/Themed';
import { msDifferenceToCounter } from '../../utils/functions';
import { Spacer } from '../Spacer';

interface props {
    duration?: number,
    overallDuration?: number,
    isFirstLap: boolean,
}

export const CurrentLapTicker: React.FC<props> = ({ duration, overallDuration, isFirstLap }) => {
    return (
        <View style={styles.container}>
            {(duration && duration > 0) ? (
                <Text style={styles.countText}>
                    { msDifferenceToCounter(duration, true)}
                </Text>
            ) : (
                <Text style={styles.countText}>
                    00:00
                </Text>
            )}
            {(isFirstLap && overallDuration && overallDuration > 0) ? (<>
                <Spacer
                    height={2}
                    width={150}
                    color='lightgray'
                />
                <Text style={styles.countText}>
                    {msDifferenceToCounter(overallDuration, true)}
                </Text>
            </>) : null}
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: 40,

    },
    countText: {
        fontSize: 40,
        fontFamily: "Futura",
    },
});
