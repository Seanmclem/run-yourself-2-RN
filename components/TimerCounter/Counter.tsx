import React, { useEffect, useRef, useState } from 'react';
import { AppState, AppStateStatus, StyleSheet } from 'react-native';

import { runHistory, getData } from '../../stores/asyncStore'

import { View } from '../../components/Themed';
import { ButtonArea } from './ButtonArea'
import { CurrentLapTicker } from './CurrentLapTicker';
import { LapsList } from './LapsList';

// 10 digit timestamp is milliseconds aka unix_timestamp
// 

import { storeData } from '../../stores/asyncStore'

export interface Lap {
    start: number,
    end?: number,
    duration?: number,
}

export interface Run {
    overallStart: number,
    overallEnd?: number,
    overallDuration?: number;
    laps: Lap[],
    currentLapIndex: number
    isDone: boolean,
}

export const Counter = () => {
    const [_, setCurrentTime] = useState<number>() // ticker for state updates

    const [run, setRun] = useState<Run>()
    const [diffInterval, setDiffInterval] = useState<any>(undefined)

    const currentLapRef = useRef<Lap>({
        start: 0
    })

    useEffect(() => {
        // AppState.addEventListener("change", _handleAppStateChange);

        return () => {
            AppState.removeEventListener("change", _handleAppStateChange);
        };
    }, []);



    const getNowTimestamp = () => (new Date()).getTime()

    const startTimer = (resume?: boolean) => {
        // if (!resume) {
        const newTime = getNowTimestamp()
        currentLapRef.current = { start: newTime }

        const newRun: Run = {
            overallStart: newTime,
            laps: [],
            currentLapIndex: 0,
            isDone: false
        }
        setRun(newRun)
        // }

        setDiffInterval(setInterval(looper, 50)) // TODO: make a ref..
    }

    const looper = () => {
        const newTime = getNowTimestamp()

        if (currentLapRef) {
            const originalStart = currentLapRef.current.start;
            const newDuration = newTime - originalStart;

            currentLapRef.current = {
                start: originalStart,
                duration: newDuration,
            }
        }

        setCurrentTime(newTime)
    }

    const stopTimer = () => {
        const newRun = endLap(true)
        if (newRun) {
            storeData({ newRun })
        }

        currentLapRef.current = {
            start: 0,
            duration: 0,
        }
        clearInterval(diffInterval)
        setDiffInterval(undefined)
        setCurrentTime(undefined)
    }

    const endLap = (theEnd?: boolean) => {
        currentLapRef.current = {
            start: currentLapRef.current.start,
            end: getNowTimestamp(),
            duration: currentLapRef.current.duration,
        }
        debugger
        if (run) {
            const laps = [...run.laps, currentLapRef.current]
            const overallDuration = laps.reduce((prev, curr) => prev + (curr.duration || 0), 0)

            const newRun: Run = {
                overallStart: run.overallStart,
                laps,
                currentLapIndex: run.currentLapIndex + 1,
                overallDuration,
                isDone: !!theEnd,
            }
            // ^here, need to save this^ to state-mgr/local-storage/db

            console.log({ overallDuration })
            setRun(newRun)

            return newRun
        }
    }

    const nextLap = () => {
        endLap()
        currentLapRef.current = {
            start: getNowTimestamp(),
            duration: 0,
        }
    }


    const _handleAppStateChange = async (nextAppState: AppStateStatus) => {
        if (nextAppState === 'background') {
            console.log("App has gone Background!");

            if (run) {
                const laps = [...run.laps, currentLapRef.current]

                const newRun: Run = {
                    overallStart: run.overallStart,
                    laps,
                    currentLapIndex: run.currentLapIndex + 1,
                    overallDuration: undefined,
                    isDone: false,
                }

                runHistory.runInProgress = newRun;
            }


        }
        if (nextAppState === 'active') {
            await getData()

            if (runHistory?.runInProgress) {
                const LapsFromCurrentRun = runHistory.runInProgress.laps
                const lastOrOnlyLap = LapsFromCurrentRun[runHistory.runInProgress.currentLapIndex]

                if (lastOrOnlyLap) {
                    currentLapRef.current = { ...lastOrOnlyLap }
                    debugger;
                    runHistory.runInProgress.laps.splice(runHistory.runInProgress.currentLapIndex, 1)
                    debugger;
                    setRun(runHistory.runInProgress)
                    startTimer(true)
                }
            }


            console.log("App has come to the gone ACTIVE")
        };
    }

    const isRunning = !(currentLapRef?.current?.start === 0)
    const overallDuration = (run?.overallDuration || 0) + (currentLapRef?.current?.duration || 0)

    console.log({ overallDuration })

    return (
        <View style={styles.container}>
            <CurrentLapTicker
                duration={currentLapRef?.current?.duration}
                overallDuration={overallDuration}
                isFirstLap={!!run?.laps?.length}
            />

            <LapsList
                laps={run?.laps}
                isDone={!!run?.isDone}
                overallDuration={run?.overallDuration}
            />

            <ButtonArea
                isRunning={isRunning}
                startTimer={startTimer}
                nextLap={nextLap}
                stopTimer={stopTimer}
            />
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        margin: 30,
        flex: 1,
        alignItems: 'center',
        justifyContent: 'space-between',
    },
});
