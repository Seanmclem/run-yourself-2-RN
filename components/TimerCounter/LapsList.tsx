import * as React from 'react';
import { StyleSheet } from 'react-native';
import { Text, View } from '../../components/Themed';
import { msDifferenceToCounter } from '../../utils/functions';
import { Spacer } from '../Spacer';
import { Lap } from './Counter';

interface props {
    laps?: Lap[],
    isDone: boolean,
    overallDuration?: number,
}

export const LapsList: React.FC<props> = ({ laps, isDone, overallDuration }) => {
    return (
        <View style={styles.container}>
            {laps ? (<>
                {isDone && (<>
                    <Text style={styles.item}>
                        Run Summary:
                    </Text>
                    <Spacer
                        height={2}
                        width='100%'
                        color='gray'
                    />
                </>)}
                {laps.map((lap, index) => (
                    lap.duration && (
                        <Text key={lap.start} style={styles.item}>
                            Lap {index + 1} - {msDifferenceToCounter(lap.duration)}
                        </Text>
                    )
                ))}
                {(overallDuration && isDone) && (<>
                    <Spacer
                        height={2}
                        width='100%'
                        color='gray'
                    />
                    <Text style={[styles.item]}>
                        Overall: {msDifferenceToCounter(overallDuration)}
                    </Text>
                </>)}
            </>) : (
                null
            )}
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'flex-start',
        justifyContent: 'center',
    },
    item: {
        fontSize: 18,
        marginVertical: 5,
    },
    footer: {
    }
});
