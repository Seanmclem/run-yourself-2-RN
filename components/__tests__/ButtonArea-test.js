import * as React from 'react';
import renderer from 'react-test-renderer';

import { ButtonArea } from '../TimerCounter/ButtonArea';

describe(`ButtonArea tests`, () => {
    it(`renders correctly`, () => {
        const tree = renderer.create(<ButtonArea />).toJSON();

        expect(tree).toMatchSnapshot();
    });
})


