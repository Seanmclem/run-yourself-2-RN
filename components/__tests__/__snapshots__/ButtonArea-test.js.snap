// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonArea tests renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#009688",
        "borderRadius": 10,
        "elevation": 8,
        "opacity": 1,
        "paddingHorizontal": 12,
        "paddingVertical": 10,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "center",
          "color": "#fff",
          "fontSize": 18,
          "fontWeight": "bold",
          "textTransform": "uppercase",
        }
      }
    >
      Start Timer
    </Text>
  </View>
</View>
`;
